packages/vite/src/node/optimizer/scan.ts

```diff
@@ -264,37 +262,40 @@ function esbuildScanPlugin(
               const src = srcMatch[1] || srcMatch[2] || srcMatch[3]
               js += `import ${JSON.stringify(src)}\n`
             } else if (content.trim()) {
-              // There can be module scripts (`<script context="module">` in Svelte and `<script>` in Vue)
+              // The reason why virtual modules are needed:
+              // 1. There can be module scripts (`<script context="module">` in Svelte and `<script>` in Vue)
               // or local scripts (`<script>` in Svelte and `<script setup>` in Vue)
+              // 2. There can be multiple module scripts in html
               // We need to handle these separately in case variable names are reused between them
-              const contextMatch = openTag.match(contextRE)
-              const context =
-                contextMatch &&
-                (contextMatch[1] || contextMatch[2] || contextMatch[3])
-              if (
-                (path.endsWith('.vue') && setupRE.test(openTag)) ||
-                (path.endsWith('.svelte') && context !== 'module')
-              ) {
-                // append imports in TS to prevent esbuild from removing them
-                // since they may be used in the template
-                const localContent =
-                  content +
-                  (loader.startsWith('ts') ? extractImportPaths(content) : '')
-                localScripts[path] = {
-                  loader,
-                  contents: localContent
+
+              // append imports in TS to prevent esbuild from removing them
+              // since they may be used in the template
+              const contents =
+                content +
+                (loader.startsWith('ts') ? extractImportPaths(content) : '')
+
+              const key = `${path}?id=${scriptId++}`
+
+              if (contents.includes('import.meta.glob')) {
+                scripts[key] = {
+                  // transformGlob already transforms to js
+                  loader: 'js',
+                  contents: await transformGlob(
+                    contents,
+                    path,
+                    config.root,
+                    loader
+                  )
                 }
-                js += `import ${JSON.stringify(virtualModulePrefix + path)}\n`
               } else {
-                js += content + '\n'
+                scripts[key] = {
+                  loader,
+                  contents
+                }
               }
-            }
-          }
 
-          // `<script>` in Svelte has imports that can be used in the template
-          // so we handle them here too
-          if (loader.startsWith('ts') && path.endsWith('.svelte')) {
-            js += extractImportPaths(js)
+              js += `import ${JSON.stringify(virtualModulePrefix + key)};\n`
```

Comment:
```suggestion
              js += `import ${JSON.stringify(virtualModulePrefix + key)}\n`
```

minor nitpick to be consistent with the above similar code too :)